pkg/sfu/forwarder.go

```diff
@@ -520,6 +518,156 @@ func (f *Forwarder) ProvisionalAllocate(availableChannelCapacity int64, layers V
 	return 0
 }
 
+func (f *Forwarder) ProvisionalAllocateGetCooperativeTransition() VideoTransition {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	if f.provisional.muted {
+		f.provisional.layers = InvalidLayers
+		return VideoTransition{
+			from:           f.targetLayers,
+			to:             InvalidLayers,
+			bandwidthDelta: 0 - f.lastAllocation.bandwidthRequested,
+			// LK-TODO should this take current bitrate of current target layers?
+		}
+	}
+
+	// check if we should preserve current target
+	if f.targetLayers != InvalidLayers {
```

Comment:
is the idea of this block to see if maximalLayers have changed since? (i.e. due to uptrack publishing limitations)

Reply:
Yeah

Reply:
Actually only partially correct. Sorry. This is the case where the stream is currently active, but the transition is asking for more (could be new available published layer OR it could be a higher subscribed layer too). This is preventing an up move when congested.

Reply:
I see, that makes sense.. Following the logic,  if targetLayer is equal or less than maximal, it would keep target. So it could move higher as long as it's lower than max. Am I reading this correctly?

Reply:
That is correct David. Basically, we got a callback from that track saying it needs a change. Because of implementation challenges, I do not explicitly say what is the change in the callback (BTW, that means I can probably combine three callbacks into one I think, have been thinking about it, but not making that change yet).

StreamAllocator acknowledges that request and calls this routine when in congested. This routine is the one which has full visibility into current state of track (mute/unmute, published layers changed, subscribed layers changed etc.). So, the resulting move could be up or down. As the channel is congested, stream allocator is telling this module two things
1. Do not make any up moves as that will affect channel.
2. But, if you are paused, let me know the minimal you need to start streaming so that I can try to satisfy the goal of all tracks always streaming.

I am adding comments and also making another change to take the transition in the above code if the new target is lower than the old target, i. e. it is currently streaming but could go down to a lower layer and hence save bits. No reason not to take that transition. Will put a pointer to the commit here once I am done. Hopefully soon.

Reply:
sounds good. thanks for the explanation.

Reply:
@davidzhao Here is the commit. Let me know if I can add some more clarity. Also, changed when a track is added/removed from stream allocator so that we are not doing any allocations till bound. Let me know if there is a better way to do that.